dbms/src/Core/SpillHandler.cpp

```diff
@@ -95,37 +131,69 @@ void SpillHandler::spillBlocks(const Blocks & blocks)
 
 void SpillHandler::finish()
 {
-    if (likely(writer != nullptr))
+    if (likely(current_spilled_file_index >= 0))
     {
-        auto spill_details = writer->finishWrite();
-        spilled_files[current_spilled_file_index]->updateSpillDetails(spill_details);
+        if (writer != nullptr)
+        {
+            auto spill_details = writer->finishWrite();
+            spilled_files[current_spilled_file_index]->updateSpillDetails(spill_details);
+            auto current_spill_details = spilled_files[current_spilled_file_index]->getSpillDetails();
+            if (!spiller->enable_append_write || isSpilledFileFull(current_spill_details.rows, current_spill_details.data_bytes_uncompressed))
+            {
+                /// always mark full if enable_append_write is false here, since if enable_append_write is false, all the files are treated as full file
+                spilled_files[current_spilled_file_index]->markFull();
+            }
+        }
+        else
+        {
+            /// writer == nullptr means no write to current file
+            if (!current_append_write)
+            {
+                /// if current_append_write is true, we still need to put the original file back, so can not discard the file here
+                --current_spilled_file_index;
+                spilled_files.pop_back();
+            }
```

Comment:
It does not seem necessary to deal with this case, because no new file is created without writing to the block.
And `current_append_write` is useless now?

Reply:
Yes, I will remove it